src/io/file_factory.rs

```diff
@@ -0,0 +1,192 @@
+
+        let file_format = DataFileFormat::from_str(
+            &table_metatdata
+                .properties
+                .as_ref()
+                .and_then(|prop| prop.get(DEFAULT_FILE_FORMAT))
+                .cloned()
+                .unwrap_or(DEFAULT_FILE_FORMAT_DEFAULT.to_string()),
+        )
+        .ok();
+
+        let data_location = if let Some(prop) = &table_metatdata.properties {
+            if let Some(value) = prop.get(WRITE_DATA_LOCATION) {
+                value.clone()
+            } else if let Some(value) = prop.get(WRITE_FOLDER_STORAGE_LOCATION) {
+                value.clone()
+            } else {
+                format!("{}/data", &table_metatdata.location)
+            }
+        } else {
+            format!("{}/data", &table_metatdata.location)
+        };
+
+        Self {
+            operator,
+            partition_id,
+            task_id,
+            suffix: "".to_string(),
+            file_format,
+            data_location,
+            operation_id,
+        }
+    }
+
+    /// Provide a suffix for the output file.
+    pub fn with_suffix(mut self, suffix: String) -> Self {
+        self.suffix = suffix;
+        self
+    }
+
+    /// Provide a output file format to override default file format.
+    ///
+    /// The default file format will try to parse from table properties and if not found, it will use parquet.
+    pub fn with_format(mut self, format: DataFileFormat) -> Self {
+        self.file_format = Some(format);
+        self
+    }
+
+    /// Consumes the builder to build a new OutputFileFactory.
+    pub fn try_build(self) -> Result<OutputFileFactory> {
+        Ok(OutputFileFactory {
+            operator: self.operator,
+            partition_id: self.partition_id,
+            task_id: self.task_id,
+            suffix: self.suffix,
+            operation_id: self.operation_id,
+            file_format: self.file_format.ok_or(error::Error::new(
+                crate::ErrorKind::IcebergDataInvalid,
+                "ouput file format is invalid",
+            ))?,
+            data_location: self.data_location,
+            file_count: AtomicUsize::new(1),
+        })
+    }
+}
+
+/// OutputWriterFactory responsible for generating a writer for a unique but recognizable data/delete file.
+///
+/// # TODO
+///
+/// Only support unpartitioned write now.
+pub struct OutputFileFactory {
+    operator: Operator,
+    partition_id: usize,
+    task_id: usize,
+    operation_id: String,
+    file_format: DataFileFormat,
+    suffix: String,
+    /// # TODO
+    ///
+    /// May need a LocationProvider trait here to support custom location. It's useful in some case: https://github.com/apache/iceberg/issues/1911.
+    data_location: String,
+    file_count: AtomicUsize,
+}
+
+impl OutputFileFactory {
+    /// Create a new writer for unpartitioned write. Return the writer and the file name.
+    pub async fn try_new_unpartitioned_writer(&self) -> Result<(Writer, String)> {
```

Comment:
What doesn the `unpartitioned` mean? Seems we already has `partition_id` here.

Reply:
The concept of this partition is the partition in iceberg. You can compare it with the interface to create partition write. 
https://github.com/apache/iceberg/blob/0c37a971430807a329adff81ebc7bbe6ee0020bf/core/src/main/java/org/apache/iceberg/io/OutputFileFactory.java#L114

And partition_id seems unrelated for it, it's just a id provide by the writer?
https://github.com/apache/iceberg/blob/0c37a971430807a329adff81ebc7bbe6ee0020bf/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java#L341